lxd/storage/backend_lxd.go

```diff
@@ -1372,13 +1366,21 @@ func (b *lxdBackend) RefreshCustomVolume(projectName string, srcProjectName stri
 
 			revert.Add(func() { _ = VolumeDBDelete(b, projectName, newSnapshotName, vol.Type()) })
 
-			// Extend the list of target snaphots to not require loading all of them again from DB.
-			targetSnapshots = append(targetSnapshots, targetSnapVol)
-
 			// Extend the list of snapshots that require refresh.
 			srcSnapVols = append(srcSnapVols, srcSnap.Name)
 		}
 
+		targetSnaps, err := VolumeDBSnapshotsGet(b, projectName, volName, drivers.VolumeTypeCustom)
```

Comment:
can we retrieve the snapshots list once and then remove the items that we delete before using the remaining items here?

Reply:
Additionally we can check based on the snapshots creation date to insert it at the right location in the `[]drivers.Volume` slice. Since the `drivers.Volume` doesn't have the creation date, the logic needs to be changed a bit. I'll propose another change and mark is as draft for now.

Reply:
Well im not sure about the creation date - the DB order is the most important one, so dont worry too much about creation date when making the snapshot order.

Reply:
When fetched from the DB the snapshots are sorted by their creation date. If we recycle the list, after removing the snapshots that have been deleted on the source volume, we have to insert new snapshots at their exact location as they would be returned from the DB.

As an example if I remove snap1 out of [snap0, snap1, snap2] and refresh the target volume, snap1 cannot be just appended at the end of the list since it's actual creation date on the source volume was before snap2 was created.
When we create snapshots on the target we also pick the creation date from the source.

I have updated it now to work with a single DB query using the new `slices.*` methods.

Reply:
See https://github.com/canonical/lxd/pull/12983#issuecomment-1971306009.